adr/0011-sap-system-database-aggregate-split.md

```diff
@@ -0,0 +1,33 @@
+# 11. SAP system and database aggregate split
+
+Date: 2024-04-04
+
+## Status
+
+Accepted
+
+## Context
+
+In the initial implementation of Trento, the SAP system and database states are stored in the same `sap_system.ex` aggregate. This is making the assumption that the relationship between a SAP system and a database is of 1 to 1, meaning that a database is only used by one SAP system. This is not an accurate description of the reality, as a database can handle multiple SAP systems in a multi-tenant environment.
+
+In this scenario, when a multi-tenant database is discovered, it only can manage a unique SAP system and its application instances, making the multi-tenant scenario not supported.
+
+This must be reverted, as a multi-tenant scenario is pretty common, and many deployments have them.
+
+## Decision
+
+The currently existing `sap_system.ex` aggregate will be separated in 2, creating a new `database.ex` aggregate that will handle databases individually. This way, the lifecycle of a database is extracted from the SAP system one, making them independent. Having this, a new 1 to N relationship will be created between the database and SAP system aggregates (read models being more precise). Now on, the SAP system ID will be calculated combining the database ID and its tenant name, creating a new and unique ID for each SAP system.
+
+The relationship between database and SAP systems will continue existing in many senses, and most of the related events will be handled by event handlers. For example, when a database is deregistered, or its health updated, these events will be listened, and once they happen, they will affect the SAP system aggregate via commands. At the end, both aggregates continue being pure, and connected through this kind of mechanisms.
```

Comment:
> and most of the related events will be handled by event handlers.

Does this mean the messages will still be handled by the same event handlers, effectively sharing some handlers?

Reply:
Not really. It means that the logic behing this connection between SAP system and database that until that was handled in the SAP system aggregate, now if handled by event handlers.
At the end, when both were 1 aggregate, you can do everything in the aggregate code, but once they are different things, "external" side effect like event handlers are required